File: scripts/validate_observability_alertmanager.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "yaml"
require "base64"

EXPECTED_SECRET = "alertmanager-pagerduty"
EXPECTED_FILE = "/etc/alertmanager/secrets/#{EXPECTED_SECRET}/routing-key"
EXPECTED_MATCHERS = [
  'alertname="SugarkubePagerDutyTest"',
  'environment="staging"',
  'cluster="sugarkube-int"',
  'severity="critical"'
].freeze

def fail_validation(message)
  warn "ERROR: invalid staging Alertmanager configuration: #{message}"
  exit 1
end

config_only = %w[--config --base64-config].include?(ARGV.first)
base64_config = ARGV.first == "--base64-config"
ARGV.shift if config_only
path = ARGV.fetch(0) { fail_validation("manifest or configuration path is required") }
input = path == "-" ? $stdin.read : File.read(path)
if config_only
  input = Base64.strict_decode64(input) if base64_config
  config = YAML.safe_load(input, permitted_classes: [], aliases: false)
else
  documents = YAML.load_stream(input).compact
  alertmanager = documents.find do |document|
    document.is_a?(Hash) && document["kind"] == "Alertmanager" &&
      document.dig("metadata", "name") == "kube-prometheus-stack-alertmanager"
  end
  fail_validation("expected Alertmanager custom resource is missing") unless alertmanager

  secrets = alertmanager.dig("spec", "secrets")
  fail_validation("Alertmanager must reference only #{EXPECTED_SECRET}") unless secrets == [EXPECTED_SECRET]

  config_secret = documents.find do |document|
    document.is_a?(Hash) && document["kind"] == "Secret" &&
      document.dig("metadata", "name") == "alertmanager-kube-prometheus-stack-alertmanager"
  end
  fail_validation("chart-rendered Alertmanager configuration Secret is missing") unless config_secret

  raw_config = config_secret.dig("stringData", "alertmanager.yaml")
  raw_config ||= Base64.strict_decode64(config_secret.dig("data", "alertmanager.yaml").to_s)
  fail_validation("Alertmanager configuration is missing") if raw_config.empty?
  config = YAML.safe_load(raw_config, permitted_classes: [], aliases: false)
end
fail_validation('root receiver must remain "null"') unless config.dig("route", "receiver") == "null"

routes = config.dig("route", "routes")
pagerduty_routes = Array(routes).select { |route| route.is_a?(Hash) && route["receiver"] == "pagerduty-synthetic" }
unless pagerduty_routes.length == 1 && pagerduty_routes.first["matchers"] == EXPECTED_MATCHERS
  fail_validation("PagerDuty must have exactly the narrow synthetic route")
end

receivers = config["receivers"]
null_receivers = Array(receivers).select { |receiver| receiver == {"name" => "null"} }
pagerduty_receivers = Array(receivers).select do |receiver|
  receiver.is_a?(Hash) && receiver["name"] == "pagerduty-synthetic"
end
unless null_receivers.length == 1 && pagerduty_receivers.length == 1 && receivers.length == 2
  fail_validation("only the null and synthetic PagerDuty receivers are allowed")
end

pagerduty_configs = pagerduty_receivers.first["pagerduty_configs"]
unless pagerduty_configs == [{"routing_key_file" => EXPECTED_FILE, "send_resolved" => true}]
  fail_validation("PagerDuty must use the expected routing-key file with send_resolved enabled")
end

forbidden = %w[routing_key service_key]
walk = lambda do |value|
  case value
  when Hash
    fail_validation("inline PagerDuty credential fields are forbidden") unless (value.keys & forbidden).empty?
    value.each_value { |child| walk.call(child) }
  when Array
    value.each { |child| walk.call(child) }
  end
end
walk.call(config)

puts "Rendered Alertmanager route, receiver, Secret mount contract, and file reference are valid."

File: scripts/validate_observability_alertmanager.py
#!/usr/bin/env python3
"""Validate the staging Alertmanager render without revealing secret material."""

import base64
import json
import subprocess
import sys
from pathlib import Path

SECRET = "alertmanager-pagerduty"
RECEIVER = "pagerduty-synthetic"
FILE = f"/etc/alertmanager/secrets/{SECRET}/routing-key"
MATCHERS = {
    'alertname="SugarkubePagerDutyTest"',
    'environment="staging"',
    'cluster="sugarkube-int"',
    'severity="critical"',
}


def load_documents(path: Path) -> list[dict]:
    ruby = subprocess.run(
        [
            "ruby",
            "-ryaml",
            "-rjson",
            "-e",
            "puts JSON.generate(YAML.load_stream(File.read(ARGV[0])))",
            str(path),
        ],
        check=True,
        capture_output=True,
        text=True,
    )
    documents = json.loads(ruby.stdout)
    return [item for item in documents if isinstance(item, dict)]


def validate_config(config: dict) -> None:
    route = config.get("route", {})
    if route.get("receiver") != "null":
        raise ValueError('root receiver must remain "null"')
    routes = route.get("routes", [])
    pd_routes = [item for item in routes if item.get("receiver") == RECEIVER]
    if len(pd_routes) != 1 or set(pd_routes[0].get("matchers", [])) != MATCHERS:
        raise ValueError("PagerDuty must have exactly the narrow synthetic route")
    receivers = [item for item in config.get("receivers", []) if item.get("name") == RECEIVER]
    if len(receivers) != 1 or len(receivers[0].get("pagerduty_configs", [])) != 1:
        raise ValueError("expected exactly one synthetic PagerDuty receiver")
    pagerduty = receivers[0]["pagerduty_configs"][0]
    if pagerduty.get("routing_key_file") != FILE or pagerduty.get("send_resolved") is not True:
        raise ValueError("PagerDuty receiver requires the expected file and send_resolved")
    if "routing_key" in pagerduty or "service_key" in pagerduty:
        raise ValueError("inline PagerDuty credentials are forbidden")


def validate_render(path: Path) -> None:
    documents = load_documents(path)
    alertmanagers = [doc for doc in documents if doc.get("kind") == "Alertmanager"]
    if len(alertmanagers) != 1 or alertmanagers[0].get("spec", {}).get("secrets") != [SECRET]:
        raise ValueError(f"Alertmanager must reference Secret {SECRET}")
    configs = [
        doc
        for doc in documents
        if doc.get("kind") == "Secret"
        and doc.get("metadata", {}).get("name") == "alertmanager-kube-prometheus-stack-alertmanager"
    ]
    if len(configs) != 1:
        raise ValueError("rendered Alertmanager configuration Secret is missing")
    encoded = configs[0].get("data", {}).get("alertmanager.yaml")
    if not isinstance(encoded, str):
        raise ValueError("rendered Alertmanager configuration is missing")
    raw = base64.b64decode(encoded, validate=True)
    parsed = subprocess.run(
        [
            "ruby",
            "-ryaml",
            "-rjson",
            "-e",
            "puts JSON.generate(YAML.safe_load(STDIN.read, aliases: false))",
        ],
        input=raw,
        check=True,
        capture_output=True,
    )
    validate_config(json.loads(parsed.stdout))


def validate_secret_json(path: Path) -> None:
    secret = json.loads(path.read_text(encoding="utf-8"))
    encoded = secret.get("data", {}).get("alertmanager.yaml")
    if not isinstance(encoded, str):
        raise ValueError("loaded Alertmanager configuration is missing")
    raw = base64.b64decode(encoded, validate=True)
    parsed = subprocess.run(
        [
            "ruby",
            "-ryaml",
            "-rjson",
            "-e",
            "puts JSON.generate(YAML.safe_load(STDIN.read, aliases: false))",
        ],
        input=raw,
        check=True,
        capture_output=True,
    )
    validate_config(json.loads(parsed.stdout))


if __name__ == "__main__":
    try:
        if len(sys.argv) == 2:
            validate_render(Path(sys.argv[1]))
        elif len(sys.argv) == 3 and sys.argv[1] == "--config-secret-json":
            validate_secret_json(Path(sys.argv[2]))
        else:
            raise ValueError(
                "usage: validate_observability_alertmanager.py [--config-secret-json] FILE"
            )
    except (ValueError, OSError, subprocess.SubprocessError, json.JSONDecodeError) as exc:
        print(f"ERROR: invalid Alertmanager structure: {exc}", file=sys.stderr)
        raise SystemExit(1)
    print("Alertmanager PagerDuty structure validated (file reference only).")
